### DISCUSSION TITLE
Display measurement progress in PyQt progressbar

### DISCUSSION THREAD

**Discussion by S-M-V** [microsoft/Qcodes#3805] [Q&A]
Hello,

I'm new in QCodes, and I would like to add a small progress bar in a simple Dialog, to show the user the measurent progress.
For doing that I have done a QThread, and inside I'm running the following code (from the 15 minutes to QCoDes):
```
class myThread(QThread):
    def run(self):
      context_meas.write_period = 2

      point=0
      with context_meas.run() as datasaver:
          for set_v in np.linspace(0, 25, 10):
              dac.ch1.set(set_v)
              get_v = dmm.v1.get()
              datasaver.add_result((dac.ch1, set_v),
                                   (dmm.v1, get_v))
              self.update_progress.emit(100*point/len(10))
              point+=1
          # Convenient to have for plotting and data access
          dataset = datasaver.dataset
```

When running I got the following error message:
```
sqlite3.ProgrammingError: SQLite objects created in a thread can only be used in that same thread. The object was created in thread id 50284 and this is thread id 36852.
```

Is there a simple workarround to do this?

And more generaly, which is the proper way to deal with threading in QCoDes?

Thank you in advance.

Regards,
Samuel.

**Comment by jenshnielsen** [comment]
For a progessbar I would strongly recommend that you take a look at tqdm or something similar rather than a qtthread. That is actually already build in to the dond functions here e.g. here https://github.com/QCoDeS/Qcodes/blob/master/qcodes/utils/dataset/doNd.py#L638

**Reply by jenshnielsen** [reply]
Concerning threading that depends a bit on what you want to do. 

QCoDeS has support for writing data in the background with a flag to the measurement runner 
https://qcodes.github.io/Qcodes/examples/DataSet/Saving_data_in_the_background.html

And there is support for reading from instruments in different threads 
https://qcodes.github.io/Qcodes/examples/DataSet/Threaded%20data%20acquisition.html

**Reply by S-M-V** [reply]
Thank you very much for your fast answer!

I will work following your suggestions with the dond rather than with "for cycles" (I am still exploring all options of QCoDes so sorry for dummy questions).

Thanks a lot,

Samuel.
